### ISSUE TITLE
superclass mismatch for class MysqlChefGem

### ISSUE THREAD

**Issue by ikozel** [phlipper/chef-percona#266]
I'm using database cookbook: https://github.com/opscode-cookbooks/database

And running into this error: 

Compiling Cookbooks...
       
       ================================================================================
       Recipe Compile Error in /tmp/kitchen/cache/cookbooks/percona/libraries/provider_mysql_chef_gem.rb
       ================================================================================
       
       TypeError
       ---------
       superclass mismatch for class MysqlChefGem
       
       Cookbook Trace:
       ---------------
         /tmp/kitchen/cache/cookbooks/percona/libraries/provider_mysql_chef_gem.rb:10:in `<class:Provider>'
         /tmp/kitchen/cache/cookbooks/percona/libraries/provider_mysql_chef_gem.rb:2:in `<class:Chef>'
         /tmp/kitchen/cache/cookbooks/percona/libraries/provider_mysql_chef_gem.rb:1:in `<top (required)>'
       
       Relevant File Content:
       ----------------------
       /tmp/kitchen/cache/cookbooks/percona/libraries/provider_mysql_chef_gem.rb:
       
         3:      # rubocop:disable LineLength
         4:      #
         5:      # Public:
         6:      # Monkey patch to not install mysql client dev libraries over ours
         7:      # https://github.com/opscode-cookbooks/mysql/blob/master/libraries/provider_mysql_client_ubuntu.rb
         8:      #
         9:      # rubocop:enable LineLength
        10>>     class MysqlChefGem < Chef::Provider::LWRPBase
        11:        def action_install
        12:          converge_by "install mysql chef_gem and dependencies" do
        13:            recipe_eval do
        14:              run_context.include_recipe "build-essential::default"
        15:              run_context.include_recipe "percona::client"
        16:            end
        17:  
        18:            chef_gem "mysql" do
        19:              action :install
       
       
       Running handlers:
       [2015-03-08T01:16:02+00:00] ERROR: Running exception handlers

Any idea why would this happen and how to address it?

**Comment by ikozel** [comment]
Any suggestion on how can mysql2_chef_gem can be used with percona v0.15.5? 

cookbook 'percona', git: 'https://github.com/phlipper/chef-percona.git', tag: 'v0.15.5'

**Comment by adamkrone** [comment]
I ran into this about month ago, but I didn't have enough time to figure everything out, so I'll just post what I found.

It looks like both mysql_chef_gem and mysql2_chef_gem use nested classes to implement the mysql/mariadb providers:

https://github.com/opscode-cookbooks/mysql_chef_gem/blob/master/libraries/provider_mysql_chef_gem_mysql.rb#L4

https://github.com/sinfomicien/mysql2_chef_gem/blob/master/libraries/provider_mysql2_chef_gem_mysql.rb#L4

This avoids monkey patching, as you can just target the provider when using it:

```ruby
mysql2_chef_gem 'default' do
  provider Chef::Provider::Mysql2ChefGem::MyCustomProvider
  action :install
end
```

So, MysqlChefGem and Mysql2ChefGem aren't inheriting from anything, whereas in this cookbook they are inheriting from Chef::Provider::LWRPBase, which explains the superclass mismatch:

https://github.com/phlipper/chef-percona/blob/master/libraries/provider_mysql2_chef_gem.rb#L4

https://github.com/phlipper/chef-percona/blob/master/libraries/provider_mysql_chef_gem.rb#L10

I remember running into another issue after changing the provider, but that may have been unrelated. Hopefully this helps.

**Comment by ikozel** [comment]
@adamkrone thanks for sharing, it's was helpful. At the end I've decided to manage dB changes not from  MySQL localhostl resource, which solved the MysqlChefGem problem for me.

**Comment by phlipper** [comment]
This is resolved via #297.
